### PR TITLE
Speedup `_fmpz_vec_content`, add `_fmpz_vec_set_str` and speedup `fmpz_mat_content`

### PR DESCRIPTION
**Pull request by albinahlback**
- Add `_fmpz_vec_set_str` to set an array of `fmpz` by strings. (Implemented in order to ease the error checking for `_fmpz_vec_content`, thought it might be of use here as well.)
- Speedup `_fmpz_vec_content` mainly by utilizing the inequality `gcd(a_1, a_2, ..., a_n) ≤ gcd(a_2, ..., a_n)` and calling GMP directly instead of calling `fmpz_gcd[3]`. See profiling result below.
- Use `_fmpz_vec_content` in order to speedup `fmpz_mat_content`. Since this code is trivial, I have inlined it.

Result from `_fmpz_vec_content`:
```
  5 bits:   (min) 1.15x speedup   (max) 1.01x speedup
 10 bits:   (min) 1.10x speedup   (max) 1.10x speedup
 15 bits:   (min) 1.08x speedup   (max) 1.08x speedup
 20 bits:   (min) 1.10x speedup   (max) 1.10x speedup
 25 bits:   (min) 1.13x speedup   (max) 1.12x speedup
 30 bits:   (min) 1.15x speedup   (max) 1.15x speedup
 35 bits:   (min) 1.19x speedup   (max) 1.20x speedup
 40 bits:   (min) 1.21x speedup   (max) 1.21x speedup
 45 bits:   (min) 1.21x speedup   (max) 1.21x speedup
 50 bits:   (min) 1.22x speedup   (max) 1.23x speedup
 55 bits:   (min) 1.22x speedup   (max) 1.23x speedup
 60 bits:   (min) 1.23x speedup   (max) 1.23x speedup
 65 bits:   (min) 1.24x speedup   (max) 1.24x speedup
 70 bits:   (min) 1.23x speedup   (max) 1.23x speedup
 75 bits:   (min) 1.21x speedup   (max) 1.21x speedup
 80 bits:   (min) 1.20x speedup   (max) 1.20x speedup
 85 bits:   (min) 1.18x speedup   (max) 1.18x speedup
 90 bits:   (min) 1.17x speedup   (max) 1.17x speedup
 95 bits:   (min) 1.16x speedup   (max) 1.16x speedup
100 bits:   (min) 1.15x speedup   (max) 1.14x speedup
105 bits:   (min) 1.15x speedup   (max) 1.20x speedup
110 bits:   (min) 1.14x speedup   (max) 1.14x speedup
115 bits:   (min) 1.13x speedup   (max) 1.14x speedup
120 bits:   (min) 1.13x speedup   (max) 1.14x speedup
125 bits:   (min) 1.13x speedup   (max) 1.12x speedup
130 bits:   (min) 1.12x speedup   (max) 1.12x speedup
135 bits:   (min) 1.12x speedup   (max) 1.12x speedup
140 bits:   (min) 1.11x speedup   (max) 1.11x speedup
145 bits:   (min) 1.11x speedup   (max) 1.11x speedup
150 bits:   (min) 1.10x speedup   (max) 1.10x speedup
```
Very nice speedup!

### CODE REVIEW

**Comment by albinahlback** [comment]
> I think the failing tests are due to an issue with the profiling code.
> 
> I didn't spot an error in the implementation. I thought I did a few times, but I don't see it. You might have to trace a failing case through by hand.

Thanks for your input, Bill. I think I spotted the error.

**Comment by albinahlback** [comment]
Since `fmpz_mat_content` was inlined, should the credits go somewhere in the repo or should they be discarded? I suppose `fmpz_mat.h` is too large of a file to take some credits for.

**Comment by fredrik-johansson** [comment]
This fmpz_mat_content will not work with window matrices.

**Comment by albinahlback** [comment]
> This fmpz_mat_content will not work with window matrices.

Okay, then it's probably better to use `_fmpz_vec_content_chained`. Will speed this up first then.

**Comment by wbhart** [comment]
I think .h files are supposed to be interface files and interfaces aren't supposed to be copyrightable, so I just wouldn't put anything in there that you want credit for. It's not a good idea to be inlining very large functions anyway. It just bloats binaries.

**Comment by albinahlback** [comment]
> I think .h files are supposed to be interface files and interfaces aren't supposed to be copyrightable, so I just wouldn't put anything in there that you want credit for. It's not a good idea to be inlining very large functions anyway. It just bloats binaries.

Yeah, I will not inline it now, since it will not be as simple.

**Comment by albinahlback** [comment]
Thanks. Yeah, I should definitely search for small fmpz's from the tail as well. And yes, the code is hard to follow. I will try to put them in different function bodies instead.
